src/core/server/rendering/bootstrap/register_bootstrap_route.ts

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { IRouter } from '../../http';
+import type { BootstrapRenderer } from './bootstrap_renderer';
+
+export const registerBootstrapRoute = ({
+  router,
+  renderer,
+}: {
+  router: IRouter;
+  renderer: BootstrapRenderer;
+}) => {
+  router.get(
+    {
+      path: '/bootstrap.js',
+      options: {
+        authRequired: 'optional',
```

Comment:
@azasypkin,

This PR moves the `/bootstrap.js` endpoint out of legacy and into `core` -- as part of this change, the route went from [hapi's authMode](https://hapi.dev/api/?v=17.9.3#-routeoptionsauthmode) of `try` to `optional`.

This caused a lot of x-pack test failures, because some test suites ended up loading Kibana with an incorrect `Authorization` header (invalid credentials). In the legacy world where this used `try`, the invalid credentials allowed the `/bootstrap.js` request to succeed. Now that this route is using `authMode: 'optional'`, the request fails when presented with invalid credentials.

We previously asked the core team to implement `authMode: 'optional'` for the KP, because we needed to detect if the user had a valid session or not, without directly relying on the existence of the session cookie (https://github.com/elastic/kibana/pull/58589). Now that we have server-side sessions, it's not clear to me if we need to keep these routes as `optional`, or if we can move them to `try`.

My gut feeling is that we don't need the behavior that `optional` gives us anymore, but I wanted to double check with you, since you're well-versed on the different auth flows and states.

Reply:
I went with adding the `try` auth mode for now.

Reply:
Sorry for the delayed reply, just caught up on the emails after PTO.

> This caused a lot of x-pack test failures, because some test suites ended up loading Kibana with an incorrect Authorization header (invalid credentials).

Interesting, why do we do that? Is there any legitimate case when it makes sense?

> We previously asked the core team to implement authMode: 'optional' for the KP, because we needed to detect if the user had a valid session or not, without directly relying on the existence of the session cookie (#58589).

Right, we wanted to abstract away from the fact that session is stored in the cookie running the Core authentication hook instead. But the hook shouldn't fail, or redirect to IdP (for SAML/OIDC), or automatically initiate new session (PKI/Kerberos) if credentials aren't provided.

>  Now that we have server-side sessions, it's not clear to me if we need to keep these routes as optional, or if we can move them to try.

I'm not sure how server-side sessions would change anything here to be honest, can you please elaborate a bit?

At that time we picked `optional` as seemingly the safest option (if you send any credentials you likely expect them to be respected) and the one with the behavior similar to the ES behavior when anonymous access is enabled there (requests without credentials are treated as anonymous and requests with invalid credentials are rejected). I still feel that `try` behavior is a bit confusing (`any request credentials are attempted authentication, but if the credentials are invalid, the request proceeds regardless of the authentication error`).

Having said that I believe we can safely switch our routes (`login`, `logged_out` etc.) to `try` since our authentication pipeline only cares if authentication is strictly required or not. So we won't need `optional` anymore.